.github/workflows/ci.yml

```diff
@@ -12,98 +12,144 @@ concurrency:
+      - if: "failure()"
+        uses: "andymckay/cancel-action@271cfbfa11ca9222f7be99a47e8f929574549e0a" # v0.4
+
+  machete:
+    name: Check unused dependencies
     runs-on: ubuntu-latest
     container:
       image: ${{ vars.CI_IMAGE }}
-      # options: --privileged
     steps:
-      - uses: actions/checkout@v3
+      - name: Checkout sources
+        uses: actions/checkout@v4
+
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@23bce251a8cd2ffc3c1075eaa2367cf899916d84 # v2.7.3
+        with:
+          cache-on-failure: true
+          cache-all-crates: true
+
+      - name: Install cargo-machete
+        run: cargo install cargo-machete
+
+      - name: Check unused dependencies 
+        run: cargo machete
 
-      # - name: Install Rust stable toolchain
-      #   uses: actions-rs/toolchain@v1.0.7
-      #   with:
-      #     profile: minimal
-      #     toolchain: 1.75.0
-      #     override: true
+      - if: "failure()"
+        uses: "andymckay/cancel-action@271cfbfa11ca9222f7be99a47e8f929574549e0a" # v0.4
+
+  check:
+    name: Cargo check
+    runs-on: ubuntu-latest
+    container:
+      image: ${{ vars.CI_IMAGE }}
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v4
 
       - name: Rust Cache
-        uses: Swatinem/rust-cache@v2.5.0
+        uses: Swatinem/rust-cache@23bce251a8cd2ffc3c1075eaa2367cf899916d84 # v2.7.3
         with:
           cache-on-failure: true
           cache-all-crates: true
 
-      - name: test
-        run: |
-          # sudo apt-get update && sudo apt-get install -y iproute2 protobuf-compiler
-          echo "Installing iproute2..."
-          apt-get update && apt-get install -y iproute2 iputils-ping
-          # ip a
-          # echo "Enabling loopback interface explicitly..."
-          # ip link set lo up
-          echo "Result"
-          ip a
-          ping -c 1 127.0.0.1
-          echo "Starting a simple HTTP server..."
-          python3 -m http.server --bind 127.0.0.1 8080 &
-          sleep 1
-          curl 127.0.0.1:8080
-          echo "Running tests..."
-          cargo test
+      - name: Cargo check
+        run: cargo check
+
+      - if: "failure()"
+        uses: "andymckay/cancel-action@271cfbfa11ca9222f7be99a47e8f929574549e0a" # v0.4
 
   doc:
-    name: "doc"
+    name: Check documentation
     runs-on: ubuntu-latest
     container:
       image: ${{ vars.CI_IMAGE }}
     steps:
-      - uses: actions/checkout@v3
+      - name: Checkout sources
+        uses: actions/checkout@v4
 
       - name: Rust Cache
-        uses: Swatinem/rust-cache@v2.5.0
+        uses: Swatinem/rust-cache@23bce251a8cd2ffc3c1075eaa2367cf899916d84 # v2.7.3
         with:
           cache-on-failure: true
           cache-all-crates: true
-      - name: doc
-        run: cargo doc
+      - name: Check documentation
+        run: RUSTDOCFLAGS="-D warnings -D rustdoc::broken_intra_doc_links" cargo doc --workspace --no-deps --document-private-items
```

Comment:
Haven't checked,  `--document-private-items` does not generate an error if a private element is not documented, does it?